source/agora/consensus/protocol/Nominator.d

```diff
@@ -769,16 +769,34 @@ extern(D):
         const cur_height = this.ledger.height();
         log.trace("Received BLOCK SIG {} from node {} for block {}",
                     block_sig.signature, block_sig.utxo, block_sig.height);
-        if (block_sig.height > cur_height)
-            return BlockHeader.init;
-        const block = this.ledger.getBlocksFrom(Height(block_sig.height)).front;
-        if (!this.collectBlockSignature(block_sig, block.hashFull()))
-            return BlockHeader.init;
-        BlockHeader header = block.header.clone();
-        this.updateMultiSignature(header);
-        const updated_sig = header;
-        this.ledger.updateBlockMultiSig(updated_sig);
-        return updated_sig;
+
+        const pendingBlockSig = (block_sig.height == this.pending_block.header.height);
+        log.dbg("{}: Signature is for pending block #{} still to be externalized", __FUNCTION__, this.pending_block.header.height);
+
+        if (!pendingBlockSig && block_sig.height > cur_height) // Too early for this signature
+        {
+            early_sigs.put(block_sig);
+            return BlockHeader.init; // We want the caller to ignore this signature
+        }
+
+        Block block = pendingBlockSig ? this.pending_block : this.ledger.getBlocksFrom(Height(block_sig.height)).front;
```

Comment:
Also if someone sends a signature for block `0` while we don't have a pending block signature, we will store it, won't we ?

Reply:
Added an early return for this condition of a block signature for height 0 or for the case where the signature is for more than 3 blocks into the future.